omegaconf/base.py

```diff
@@ -321,32 +306,76 @@ def _select_impl(
             throw_on_type_error=throw_on_resolution_failure,
         )
         if value is None:
-            return root, last_key, value
+            return root, last_key, None
         value = root._resolve_interpolation(
+            parent=root,
             key=last_key,
             value=value,
             throw_on_missing=False,
             throw_on_resolution_failure=True,
         )
         return root, last_key, value
 
+    def _resolve_complex_interpolation(
+        self,
+        parent: Optional["Container"],
+        value: "Node",
+        key: Any,
+        parse_tree: OmegaConfGrammarParser.ConfigValueContext,
+        throw_on_missing: bool,
+        throw_on_resolution_failure: bool,
+    ) -> Optional["Node"]:
+        """
+        A "complex" interpolation is any interpolation that cannot be handled by
+        `_resolve_simple_interpolation()`, i.e. that either contains nested
+        interpolations or is not a single "${..}" block.
+        """
+
+        from .nodes import StringNode
+
+        resolved = self.resolve_parse_tree(
+            parse_tree=parse_tree,
+            key=key,
+            parent=parent,
+            throw_on_missing=throw_on_missing,
+            throw_on_resolution_failure=throw_on_resolution_failure,
+        )
+
+        if resolved is None:
+            return None
+        elif isinstance(resolved, str):
+            # Result is a string: create a new node to store it.
+            return StringNode(
+                value=resolved,
+                key=key,
+                parent=parent,
+                is_optional=value._metadata.optional,
+            )
+        else:
+            assert isinstance(resolved, Node)
+            return resolved
+
     def _resolve_simple_interpolation(
         self,
         key: Any,
-        inter_type: str,
-        inter_key: str,
+        parent: Optional["Container"],
+        inter_type: Optional[str],
+        inter_key: Tuple[Any, ...],
```

Comment:
@omry we had an ongoing discussion regarding potentially splitting this function in two in order to avoid this weird `inter_key`: https://github.com/omry/omegaconf/pull/321#discussion_r513641106

Let me know of you believe my suggestion in that comment would make sense.

Reply:
> I would have two different functions _resolve_node_interpolation() and _resolve_resolver_interpolation() (the second name is a bit unfortunate, but it's actually the one that makes most sense IMO to be consistent with the terminology used elsewhwere)
The first one would essentially contain the code below if inter_type is None: and the second one the code below else:

Let's split to `_resolve_node_interpolation()` and `_resolve_custom_resolver()` (or maybe `_evaluate_custom_resolver()`).

Reply:
Ok I just did that -- I squashed it to keep the history clean and rebased on top of master

Reply:
Resolving that point since it's been done in the version you reviewed